osu.Game/Online/Chat/DrawableLinkCompiler.cs

```diff
@@ -1,57 +1,105 @@
 
-#nullable disable
-
+using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.Linq;
 using osu.Framework.Allocation;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Cursor;
 using osu.Framework.Graphics.Sprites;
+using osu.Framework.Graphics.UserInterface;
+using osu.Framework.Platform;
 using osu.Game.Graphics;
 using osu.Game.Graphics.Containers;
 using osu.Game.Graphics.UserInterface;
 using osu.Game.Overlays;
+using osu.Game.Overlays.OSD;
 using osuTK;
 
 namespace osu.Game.Online.Chat
 {
     /// <summary>
     /// An invisible drawable that brings multiple <see cref="Drawable"/> pieces together to form a consumable clickable link.
     /// </summary>
-    public partial class DrawableLinkCompiler : OsuHoverContainer
+    public partial class DrawableLinkCompiler : OsuHoverContainer, IHasContextMenu
     {
         /// <summary>
         /// Each word part of a chat link (split for word-wrap support).
         /// </summary>
         public readonly List<Drawable> Parts;
 
-        [Resolved(CanBeNull = true)]
-        private OverlayColourProvider overlayColourProvider { get; set; }
+        public readonly LinkDetails Link;
+
+        [Resolved]
+        private OverlayColourProvider? overlayColourProvider { get; set; }
+
+        [Resolved]
+        private OnScreenDisplay? onScreenDisplay { get; set; }
+
+        [Resolved]
+        private GameHost host { get; set; } = null!;
+
+        [Resolved]
+        private ILinkHandler? linkHandler { get; set; }
 
         public override bool ReceivePositionalInputAt(Vector2 screenSpacePos) => Parts.Any(d => d.ReceivePositionalInputAt(screenSpacePos));
 
         protected override HoverSounds CreateHoverSounds(HoverSampleSet sampleSet) => new LinkHoverSounds(sampleSet, Parts);
 
-        public DrawableLinkCompiler(ITextPart part)
-            : this(part.Drawables.OfType<SpriteText>())
+        public DrawableLinkCompiler(ITextPart part, LinkDetails link)
+            : this(part.Drawables.OfType<SpriteText>(), link)
         {
         }
 
-        public DrawableLinkCompiler(IEnumerable<Drawable> parts)
+        public DrawableLinkCompiler(IEnumerable<Drawable> parts, LinkDetails link)
         {
             Parts = parts.ToList();
+            Link = link;
         }
 
         [BackgroundDependencyLoader]
         private void load(OsuColour colours)
         {
             IdleColour = overlayColourProvider?.Light2 ?? colours.Blue;
+
+            Action ??= () =>
+            {
+                Debug.Assert(Link.Action != LinkAction.Custom);
+
+                if (linkHandler != null)
+                    linkHandler.HandleLink(Link);
+                // fallback to handle cases where OsuGame is not available, ie. tournament client.
+                else if (Link.Action == LinkAction.External)
+                    host.OpenUrlExternally(Link.Argument.ToString());
+            };
```

Comment:
This increases the scope of `DrawableLinkCompiler` a bit, but I think the name still stands? Maybe we could rename this to `DrawableLinkHandler`/`DrawableTextLinkHandler`, but I'm not sure how that covers the "effects" part of this component, so I left it as is.

Reply:
I'm going to have a rather large issue with all of this I'm afraid. Previously `DrawableLinkCompiler` was a simple wrapper for flowing text sprites to look like a link. Now it's resolving OSDs and handling clipboard and adding link-type-specific logic.

None of this should be here. It should be shipped off to a place specific to "URL-based hyperlinks".

Reply:
Hmm, you make a good point. I'll try splitting it off the compiler.